napari/utils/events/dataclass.py

```diff
@@ -223,9 +251,15 @@ def evented_post_init(self: T, *initvars) -> None:
             for em in self.events.emitters:
                 e_fields.pop(em, None)
             self.events.add(**e_fields)
+            if "values_updated" not in self.events.emitters:
+                self.events.add(values_updated=None)
         else:
             self.events = EmitterGroup(
-                source=self, auto_connect=False, **e_fields
+                source=self,
+                auto_connect=False,
+                deprecated=getattr(cls, "deprecated_events", None),
```

Comment:
I don't love adding this to each `evented_dataclass`, is there a way to add deprecated events after the object is made?

I've also found the current syntax https://github.com/napari/napari/blob/master/napari/utils/events/event.py#L692 that we added in #1582 limiting as sometimes there is not a drop in replacement provided as is, and so would probably prefer something that just let me pass the entire custom warning. Not sure if that's starting to sound like the following then https://github.com/napari/napari/blob/master/napari/utils/events/event.py#L603

Reply:
> I don't love adding this to each `evented_dataclass`, is there a way to add deprecated events after the object is made?

To achieve this `EmitterGroup` needs to be updated. 

> I've also found the current syntax [https://github.com/napari/napari/blob/master/napari/utils/events/event.py#L692](https://github.com/napari/napari/blob/master/napari/utils/events/event.py?rgh-link-date=2020-12-01T03%3A43%3A13Z#L692) that we added in #1582 limiting as sometimes there is not a drop in replacement provided as is, and so would probably prefer something that just let me pass the entire custom warning. Not sure if that's starting to sound like the following then [https://github.com/napari/napari/blob/master/napari/utils/events/event.py#L603](https://github.com/napari/napari/blob/master/napari/utils/events/event.py?rgh-link-date=2020-12-01T03%3A43%3A13Z#L603)


My idea for #1582 is to add a very simple mechanism. I think that WarningEmitter could be used in a more complicated state. But it does not support fast connection to already exists code and needs many more lines of code to be used. 

A possible refactor is to allow pass tuple (namedtuple) of arguments instead of the name of the signal, which tuple will contain information like signal name, custom error message, etc. But it will produce a more complicated code. 

As evented dataclass starts to be widely used inside napari and napari still sometimes fast changes it should start to provide a simple mechanism of backward compattybility. 

`values_updated` is part of my proposition of such API. simple deprecation method (which allows to not write many lines of code to inform the user about deprecation) also. Another option may be some __post_init__ method. but inheritance is problematic here.

Reply:
> As evented dataclass starts to be widely used inside napari and napari still sometimes fast changes it should start to provide a simple mechanism of backward compattybility.
>
> values_updated is part of my proposition of such API. simple deprecation method (which allows to not write many lines of code to inform the user about deprecation) also. Another option may be some post_init method. but inheritance is problematic here.

I think we will already need to add custom getters to do a deprecation which is very easy with the `Property` syntax, as we'll need to emit warnings when people just access the property even if they don't try and connect events to it. Given that we'll have to do custom things like that when we deprecated I'd really prefer not to add this special property "deprecated_events" to all our dataclasses right now.

I also think this is something we can solve when we deprecate our first property and not in a PR to add an `update` method to the dataclass, which is unrelated.